### PR TITLE
Howto/early stopping

### LINKED ISSUE #886: Add HOWTO: Early Stopping
Based on https://github.com/google/flax/pull/711

### PR DESCRIPTION
**Pull request by INF800**
How to documentation for early stopping closes #886

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,4 @@ build/
 .pytype
 .vscode/*
 /.devcontainer
+.venv/
```
Comment (INF800) [inline .gitignore:15]: For github codespaces

**Comment by INF800** [comment]
@marcvanzee @avital let me know if this looks good.

**Comment by avital** [comment]
Hi @INF800 -- please take a look at what the rendered doc looks lke: https://flax--1528.org.readthedocs.build/en/1528/howtos/early_stopping.html

It show a nice side-by-side diff view with the original training loop on the left and the modified version with early stopping (and everything else as well) on the right side.

Take a look at the following for an example: https://flax.readthedocs.io/en/latest/howtos/ensembling.html

Also please take a careful look at formatting and re-read it multiple times, as it stands there are some misformed sentences and the explanation can be improved.

File: .gitignore (hunk 1 above)
Comment (marcvanzee) [reply .gitignore:15]: Please add a newline

File: docs/howtos/early_stopping.rst [hunk 2]
```diff
@@ -0,0 +1,124 @@
+Early Stopping
+=============================
+Early stopping is used while building model to stop training loop when specific conditions are met. Some such conditions are:
+
+* Train loss not decreasing at all.
+* Train loss is decreasing but the decrease in loss is is insignificant.
```
Comment (marcvanzee) [inline docs/howtos/early_stopping.rst:6]:
```suggestion
- Train loss not decreasing at all.
- Train loss is decreasing but the decrease in loss is is insignificant.
```

File: docs/howtos/early_stopping.rst [hunk 3]
```diff
@@ -0,0 +1,124 @@
+Early Stopping
+=============================
+Early stopping is used while building model to stop training loop when specific conditions are met. Some such conditions are:
```
Comment (marcvanzee) [inline docs/howtos/early_stopping.rst:3]: Can you please just copy/paste a definition of early stopping? This sentence is a bit hard to parse.

File: docs/howtos/early_stopping.rst (hunk 2 above)
Comment (marcvanzee) [inline docs/howtos/early_stopping.rst:6]: The second condition repeats the first one

File: docs/howtos/early_stopping.rst [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+
+Use :code:`early_stopping.EarlyStopping` class imported from :code:`flax.training` to create an early stopping object :code:`es`.
+
+:code:`min_delta` is the minimum expected decrease in loss metric. 
+:code:`es.should_stop` becomes true if decrease in the loss metric is greater than :code:`min_delta` for consecutive :code:`patience` number of epochs. 
+Otherwise, it is :code:`false` indicating that we must continue training.
+
+
+A sample training loop looks like this
```
Comment (marcvanzee) [inline docs/howtos/early_stopping.rst:36]: Can you use syntax highlighting here for the codediff blocks like in the enselmbing HOWTO? https://github.com/google/flax/blob/main/docs/howtos/ensembling.rst

You have to append "#!" to each line you want to highlight

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Thanks for the HOWTO!
It would be great if you could add a Colab with this change as well so we can test it actually works.

Are you still planning on finishing this? (It is a while ago already since it has been created)

**Comment by INF800** [comment]
Sorry for the delay. I was under the impression that it this was merged. Will send PR by end of weekend.

**Comment by marcvanzee** [comment]
@INF800 any updates on this? Note I improved the docstring for Early Stopping in #2090 by adding a minimal example.

**Comment by marcvanzee** [comment]
@INF800 I am closing this PR due to inactivity, and like I said we have now a minimal example in our docstring already.
